Sources/BraveUI/SwiftUI/WebImageReader.swift

```diff
@@ -5,82 +5,70 @@
-  /// loaded image data, may be nil if hit from memory cache. This will only published once loading is finished
-  @Published public var isFinished: Bool = false
-  /// loading error
-  @Published public var error: Error?
-
-  private var manager = SDWebImageManager.shared
-  private var operation: SDWebImageOperation?
+public protocol WebImageDownloaderType: AnyObject {
+  func downloadImage(url: URL) async -> (UIImage?, Error?)
+  func imageFromData(data: Data) async -> UIImage?
+}
 
-  private var supportedCoders: [SDImageCoder] = [SDImageSVGNativeCoder.shared, SDImageAPNGCoder.shared, SDImageGIFCoder.shared]
-  
-  init() {}
+struct WebImageDownloaderKey: EnvironmentKey {
+  static var defaultValue: WebImageDownloaderType = SDWebImageManager.shared
+}
 
-  func load(url: URL?, options: SDWebImageOptions = []) {
-    operation = manager.loadImage(
-      with: url, options: options, progress: nil,
-      completed: { [weak self] image, data, error, _, finished, _ in
-        guard let self = self else { return }
-        self.image = image
-        self.error = error
-        if finished {
-          self.isFinished = true
-        }
-      })
+extension EnvironmentValues {
+  public var webImageDownloader: WebImageDownloaderType {
+    get { self[WebImageDownloaderKey.self] }
+    set { self[WebImageDownloaderKey.self] = newValue }
   }
+}
 
-  func cancel() {
-    operation?.cancel()
-    operation = nil
+extension SDWebImageManager: WebImageDownloaderType {
+  @MainActor public func downloadImage(url: URL) async -> (UIImage?, Error?) {
+    var operation: SDWebImageCombinedOperation?
+    return await withTaskCancellationHandler {
+      await withCheckedContinuation { continuation in
+        operation = loadImage(with: url, progress: nil) { image, _, error, _, _, _ in
+          continuation.resume(returning: (image, error))
+        }
+      }
+    } onCancel: { [operation] in
+      operation?.cancel()
+    }
   }
   
-  func load(base64Str: String, options: [SDImageCoderOption: Any] = [:]) {
-    guard base64Str.hasPrefix("data:image/") else { return }
-    guard let dataString = base64Str.separatedBy(",").last else { return }
-    
-    let data = Data(base64Encoded: dataString, options: .ignoreUnknownCharacters)
-    for coder in supportedCoders where coder.canDecode(from: data) {
-      image = coder.decodedImage(with: data, options: options)
-      break
-    }
+  public func imageFromData(data: Data) async -> UIImage? {
+    SDImageCodersManager.shared.decodedImage(with: data)
   }
 }
 
 public struct WebImageReader<Content: View>: View {
-  @StateObject private var imageManager: WalletWebImageManager = .init()
-  var url: URL?
-  var options: SDWebImageOptions
-  var coderOptions: [SDImageCoderOption: Any]
+  var url: URL
+  
+  @Environment(\.webImageDownloader) private var imageDownloader: WebImageDownloaderType
+  
+  @State private var image: UIImage?
+  @State private var error: Error?
```

Comment:
Maybe store this as `Result<UIImage, Error>` instead

Reply:
What's the reason for storing the error btw? It's not used at all anywhere in this PR

Reply:
as @StephenHeaps mentioned that `isFinished` is removed there is no way to determine image is being loaded. If both image and error are nil can be considered as its being loaded. If we don't care about loading state, i can remove it.

Reply:
Given how this was set up as a protocol, loading state should be held by `WebImageReader` prior to calling `downloadURL` then. There doesn't seem to be a need for holding onto errors though atm

Reply:
@kylehickinson i realize adding Result<UIImage, Error>/Throws is creating unnecessary complexity. also feel weird that one protocol method requires throw an error but the other just an optional UIImage. So, I'm removing the error as it doesn't seem to be worth of adding right now.

Reply:
Fine to remove the error entirely right now, there's no usage of it either way.

Reply:
`UIImage?` is now the only `State` variable being stored in [8e4439d](https://github.com/brave/brave-ios/pull/7881/commits/8e4439db703e9ca9718fade5121652e75ebc5e68)